File: src/copy-fsm/StatePattern/Parser.cs
using StatePattern.States;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace StatePattern
{
    internal class Parser
    {
        private State[] states;
        public State current { get; internal set; }
        public bool IsAcceptable { get; private set; }
        public DBTable selectedTable { get; internal set; }
        public List<string> selectedAttr { get; internal set; }
        public List<DBTable> Tables { get; private set; }

        public Parser(List<DBTable> tables)
        {
            selectedAttr = new List<string>();
            IsAcceptable = false;
            Tables = tables;
            SetStates();
            Reset();
        }

        // delegate Parse() to current state
        public string Parse(string input)
        {
            return current.Parse(this, input);
        }

        public void ChangeState(int state)
        {
           current = states[state];
           if (current is EndState)
           {
                IsAcceptable = true;
           }
        }
        
        private void SetStates()
        {
            Console.WriteLine("setting fsm states");
            states = new State[10];
            states[0] = new StartState();
            states[1] = new State1();
            states[2] = new State2();
            states[3] = new State3();
            states[4] = new State4();
            states[5] = new State5();
            states[6] = new State6();
            states[7] = new State7();
            states[8] = new EndState();
            states[9] = new ErrorState();
        }

        public void Reset()
        {
            Console.WriteLine("reseting fsm state");
            current = states[0];
            IsAcceptable = false;
        }

        internal void AddSelectedAttr(string attr)
        {
            selectedAttr.Add(attr);
        }
    }
}

File: src/js/src/Parser.ts
import { startState, state1, state2, state3, state4, state5, state6, state7, endState, errorState, IState } from "./States";

export class SQLParser {
	tokens: string[];
	attr: string[];
	states: IState[];
	state: IState;

	constructor(tokens: string[], attrs: string[]) {
		if (tokens != null && tokens.length > 0) {
			this.tokens = tokens;
		} else {
			console.error("parser error: empty token");
		}

		if (attrs != null) {
			this.attr = attrs;
		} else {
			console.log("parser error: attributes undefined");
		}

		this.states = new Array<IState>();
		this.states[0] = startState;
		this.states[1] = state1;
		this.states[2] = state2;
		this.states[3] = state3;
		this.states[4] = state4;
		this.states[5] = state5;
		this.states[6] = state6;
		this.states[7] = state7;
		this.states[8] = endState;
		this.states[9] = errorState;

		this.state = this.states[0];
	}

	parse(): void {
		for (let i: number = 0; i < this.tokens.length; i++) {
			console.log("parsing " + this.tokens[i]);
			this.state.parse(this, this.tokens[i], this.attr);
		}
	}

	changeState(stateIndex: number): void {
		this.state = this.states[stateIndex];
		console.log("new state" + stateIndex);
	}

	reset(): void {
		this.tokens = null;
		this.state = this.states[0];
	}
}